core/retries/src/main/java/software/amazon/awssdk/retries/internal/ratelimiter/RateLimiterTokenBucket.java

```diff
@@ -0,0 +1,287 @@
+ */
+
+package software.amazon.awssdk.retries.internal.ratelimiter;
+
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * The {@link RateLimiterTokenBucket} keeps track of past throttling responses and adapts to slow down the send rate to adapt to
+ * the service. It does this by suggesting a delay amount as result of a {@link #tryAcquire()} call. Callers must update its
+ * internal state by calling {@link #updateRateAfterThrottling()} when getting a throttling response or
+ * {@link #updateRateAfterSuccess()} when getting successful response.
+ *
+ * <p>The algorithm used is adapted from the network congestion avoidance algorithm
+ * <a href="https://en.wikipedia.org/wiki/CUBIC_TCP">CUBIC</a>.
+ */
+@SdkInternalApi
+public class RateLimiterTokenBucket {
+    private final AtomicReference<PersistentState> stateReference;
+    private final RateLimiterClock clock;
+
+    RateLimiterTokenBucket(RateLimiterClock clock) {
+        this.clock = clock;
+        this.stateReference = new AtomicReference<>(new PersistentState());
+    }
+
+    /**
+     * Acquires one token from the rate limiter and returns a response with a delay value (possibly zero). Callers are expected to
+     * wait for at least this amount of time before retrying the request. See {@link RateLimiterAcquireResponse#delay()}.
+     */
+    public RateLimiterAcquireResponse tryAcquire() {
+        StateUpdate<Duration> update = updateState(ts -> ts.tokenBucketAcquire(clock, 1.0));
+        return RateLimiterAcquireResponse.create(update.result);
+    }
+
+    /**
+     * Updates the estimated send rate after a throttling response.
+     */
+    public RateLimiterUpdateResponse updateRateAfterThrottling() {
+        StateUpdate<Void> update = consumeState(ts -> ts.updateClientSendingRate(clock, true));
+        return RateLimiterUpdateResponse.builder()
+            .measuredTxRate(update.newState.measuredTxRate())
+            .fillRate(update.newState.fillRate())
+            .build();
+    }
+
+    /**
+     * Updates the estimated send rate after a successful response.
+     */
+    public RateLimiterUpdateResponse updateRateAfterSuccess() {
+        StateUpdate<Void> update = consumeState(ts -> ts.updateClientSendingRate(clock, false));
+        return RateLimiterUpdateResponse.builder()
+                                        .measuredTxRate(update.newState.measuredTxRate())
+                                        .fillRate(update.newState.fillRate())
+                                        .build();
+    }
+
+    private StateUpdate<Void> consumeState(Consumer<TransientState> mutator) {
+        return updateState(ts -> {
+            mutator.accept(ts);
+            return null;
+        });
+    }
+
+    private <T> StateUpdate<T> updateState(Function<TransientState, T> mutator) {
```

Comment:
Sorry for repeating the reference of [RateLimitingTokenBucket](https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/retry/RateLimitingTokenBucket.java) but the java docs followed in that class is very helpful , can we add similar java docs here too

Reply:
Sure, we can add something similar.

Reply:
I have added some doc comments when appropriate, but I did not add any of the java docs that shows the pseudo-code of the method being implemented such as:

```
    /**
     * <pre>
     * _CUBICThrottle(rate_to_use)
     *   calculated_rate = rate_to_use * BETA
     *   return calculated_rate
     * </pre>
     */
    // Package private for testing
    double cubicThrottle(double rateToUse) {
        double calculatedRate = rateToUse * BETA;
        return calculatedRate;
    }
```

Since I think those actually makes the reading harder.

Reply:
Okay ,It would be great if the PR included a brief description of the modifications made to the classes, specifically in terms of their relationships.